cep-20.2.md

```diff
@@ -0,0 +1,829 @@
+  source_files: [glob]
+  # requirements at test time, in the target_platform architecture
+  requires: [PackageSelector]
+  # commands to execute
+  commands: [string]
+  # imports to execute with python (e.g. `import <string>`)
+  imports: [string]
+  # downstream packages that should be tested against this package
+  downstreams: [PackageSelector]
+```
+
+</details>
+
+The new test section consists of a list of test elements. Each element is executed independently and can have different requirements.
+There are multiple types of test elements defined, such as the `command` test element, the `python` test element and the `downstream` test element.
+
+Before, the test section was written to a single folder (`info/test`). In the new format, we propose to write each test element to a separate folder
+(`info/tests/<index>`). This allows us to run each test element independently.
+
+```yaml
+tests: [TestElement]
+```
+
+#### Command test element
+
+The command test element renders to a single folder with a `test_time_dependencies.json` with two keys (`build` and `run`) that contain the raw "PackageSelector" strings.
+The `script` is rendered to a `script.json` that contains the `interpreter`, `env` and other keys (as defined in the `Script` section).
+Files are copied into the `info/tests/<index>` folder.
+
+```yaml
+# script to execute
+# reuse script definition from above
+script: string | [string] | Script
+# optional extra requirements
+requirements:
+  # extra requirements with build_platform architecture (emulators, ...)
+  build: [PackageSelector]
+  # extra run dependencies
+  run: [PackageSelector]
+
+# extra files to add to the package for the test
+files:
+  # files from $SRC_DIR
+  source: [glob]
+  # files from $RECIPE_DIR
+  recipe: [glob]
+```
+
+#### Python test element
+
+The python test element renders a `test_import.py` file that contains the imports to test.
+
+```yaml
+python:
+  # list of imports to try
+  imports: [string]
+```
+
+#### Downstream test element
+
+The downstream test element renders to a `test_downstream.json` file that contains the `downstream` key with the raw "PackageSelector" string.
+
+```yaml
+downstream: PackageSelector
+```
+
+## Outputs section
+
+conda-build has very hard to understand behavior with multiple outputs. We propose some drastic simplifications.
+Each output in the new format has the same keys as a "top-level" recipe.
+
+Values from the top-level `build`, `source` and `about` sections are (deeply) merged into each of the outputs.
+
+The top-level package field is replaced by a top-level `recipe` field. The `version` from the top-level
+`recipe` is also merged into each output. The top-level name is ignored (but could be used, e.g. for
+the feedstock-name in the conda-forge case).
+
+```yaml
+# note: instead of `package`, it says `recipe` here
+recipe:
+  name: string # mostly ignored, could be used for feedstock name
+  version: string # merged into each output if not overwritten by output
```

Comment:
We also debated this. 

One alternative: keep the top-level `package` as is and rename the `package` in the outputs to `output: {name: ..., version: ...}`.

I think repeating the version for each output is somewhat annoying, unfortunately. But maybe I am overestimating how annoying it would be.

Reply:
I think that is pretty annoying. I'd be inclined to remove this, and have the version here come from the top-level package/version key. That would also be the default for any output that didn't explicitly override it. But maybe I'm just set in my ways from the current way of doing things.

Reply:
+1